cloud/pkg/apis/edgeapplication/v1alpha1/type.go

```diff
@@ -0,0 +1,221 @@
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:JSONPath=`.status.conditions[?(@.type=="Applied")].status`,name="Applied",type=string
+// +kubebuilder:printcolumn:JSONPath=`.metadata.creationTimestamp`,name="Age",type=date
+
+// EdgeApplication defines a list of resources to be deployed on the node groups.
+type EdgeApplication struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec represents the desired behavior of EdgeApplication.
+	Spec EdgeAppSpec `json:"spec"`
+
+	// Status represents the status of PropagationStatus.
+	// +optional
+	Status EdgeAppStatus `json:"status,omitempty"`
+}
+
+// EdgeAppSpec defines the desired state of EdgeApplication.
+type EdgeAppSpec struct {
+	// ResourceTemplate represents the manifest workload to be deployed on managed node groups.
+	ResourceTemplate ResourceTemplate `json:"resourceTemplate,omitempty"`
+	// WorkloadScopes represents the scope of workload to be deployed.
+	WorkloadScopes WorkloadScope `json:"workloadScopes"`
+}
+
+// WorkloadScope represents the scope of workload to be deployed.
+type WorkloadScope struct {
+	// TargetNodeGroups represents the target node groups of workload to be deployed.
+	// +optional
+	TargetNodeGroups []TargetNodeGroups `json:"targetNodeGroups,omitempty"`
+	// TargetNodes represents the target nodes of workload to be deployed.
+	// +optional
+	TargetNodes []TargetNodes `json:"targetNodes,omitempty"`
+}
+
+// TargetNodeGroups represents the target node groups of workload to be deployed.
+type TargetNodeGroups struct {
+	// Name represents the name of target node group
+	Name string `json:"name"`
+	// Overriders offers various alternatives to represent the override rules.
+	Overriders Overriders `json:"overriders,omitempty"`
+}
+
+// TargetNodes represents the target nodes of workload to be deployed.
+type TargetNodes struct {
+	// Label of target nodes
+	Label labels.Selector `json:"label"`
```

Comment:
How about supporting selecting nodes with nodename?

Reply:
Do we really need the TargetNodes field?
1. if users specify both `TargetNodeGroups` and `TargetNodes`, what will happen?
2. though we select `TargetNodes` with label, how about just creating a nodegroup for them and selecting them as in `TargetNodeGroups`?

Reply:
> 
I think it's ok with the node name.

Reply:
> Do we really need the TargetNodes field?
> 
> 1. if users specify both `TargetNodeGroups` and `TargetNodes`, what will happen?
> 2. since we select `TargetNodes` with label, how about just creating a nodegroup for them and selecting them in `TargetNodeGroups`?

1. only support one of them at the same time, if we got two specifies then an error will return? or we only support TargetNodeGroups in EdgeApplication? what's your opinion? @fisherxu @Congrool 
2. Do you mean to create a nodegroup auto?

Reply:
> Do you mean to create a nodegroup auto?

Well, I mean users can manually create a nodegroup for these nodes. Because these nodes has already had the same label which means these nodes have the same attribute, maybe location or something else, it's reasonable to organize them as a nodegroup.

Reply:
>Well, I mean users can manually create a nodegroup for these nodes. Because these nodes has already had the same label which means these nodes have the same attribute, maybe location or something else, it's reasonable to organize them as a nodegroup.

I think there're also some problems here. If we group nodes with the attribute except the location, we cannot assume that the network between these nodes is available, which means the service scope for them makes no sense. Considering this problem, should we let the users to determine which nodegroups use the service scope and which not.

Reply:
> > Well, I mean users can manually create a nodegroup for these nodes. Because these nodes has already had the same label which means these nodes have the same attribute, maybe location or something else, it's reasonable to organize them as a nodegroup.
> 
> I think there're also some problems here. If we group nodes with the attribute except the location, we cannot assume that the network between these nodes is available, which means the service scope for them makes no sense. Considering this problem, should we let the users to determine which nodegroups use the service scope and which not.

yes, the network problem really exists, would it be too complex for users to configure service scope and deploy scope both, we can add constraints descriptions for nodegroup that we only support nodes in one network to be added into the same nodegroup.

Reply:
>yes, the network problem really exists, would it be too complex for users to configure service scope and deploy scope both, we can add constraints descriptions for nodegroup that we only support nodes in one network to be added into the same nodegroup.

Agree+1
let's return back to the previous problem. Under this cercumstance, do we need to provide the service scope feature for `TargetNodes`? And users should guarrantee the network availability between these nodes selected by label.

Reply:
> > yes, the network problem really exists, would it be too complex for users to configure service scope and deploy scope both, we can add constraints descriptions for nodegroup that we only support nodes in one network to be added into the same nodegroup.
> 
> Agree+1 let's return back to the previous problem. Under this cercumstance, do we need to provide the service scope feature for `TargetNodes`? And users should guarrantee the network availability between these nodes selected by label.

no need to provide the service scope feature for `TargetNodes`, we achieve `TargetNodes` to be compatible for users deploying resources through EdgeApplication like the normal way by k8s apis .

Reply:
>no need to provide the service scope feature for TargetNodes, we achieve TargetNodes to be compatible for users deploying resources through EdgeApplication like the normal way by k8s apis .
>since we select TargetNodes with label, how about just creating a nodegroup for them and selecting them in TargetNodeGroups?

Conclude from the discussion, here's my opinion:
1. Only use `TargetNodeGroups`
2. Add a new field to NodeGroup, temporarily calling it as`EnableServiceScope`. It works like a switch to control if we need to provide the service scope function for this nodegroup. The default value is false. Users can determine to set it to true if they can guarrantee the network availability.
3. With this method, we can simplify the logic when processing the EdgeApplication.

Reply:
> > no need to provide the service scope feature for TargetNodes, we achieve TargetNodes to be compatible for users deploying resources through EdgeApplication like the normal way by k8s apis .
> > since we select TargetNodes with label, how about just creating a nodegroup for them and selecting them in TargetNodeGroups?
> 
> Conclude from the discussion, here's my opinion:
> 
> 1. Only use `TargetNodeGroups`
> 2. Add a new field to NodeGroup, temporarily calling it as`EnableServiceScope`. It works like a switch to control if we need to provide the service scope function for this nodegroup. The default value is false. Users can determine to set it to true if they can guarrantee the network availability.
> 3. With this method, we can simplify the logic when processing the EdgeApplication.

1. yes we only use `TargetNodeGroups`
2. In my opinion, we assume that the network of nodes in a node group is normal and users should guarantee this, if we add this switch and set it to 'false', users can visit the service in a  cross-partition way, it is not our intention, so the effect of the new switch is not very obvious.